veps/sig-compute/hypervisor-abstraction.md

```diff
@@ -0,0 +1,590 @@
+
+      // Override the base domain configurator
+      baseConverter.domainConfigurator = SampleDomainConfigurator{}
+      // Add an extra configurator
+      baseConverter.extraConfigurators = append(baseConverter.extraConfigurators, ExtraSampleConfigurator{})
+
+      // Custom serial attachment function
+      baseConverter.serialDeviceConverter = customSampleSerialDeviceConverter
+
+      return &SampleConverter{
+        baseConverter: baseConverter,
+      }
+    }
+    ```
+
+4. **Admission** – Create `pkg/virt-api/webhooks/validating-webhook/admitters/hypervisor/sample.go` that exports `MutateVMI` and `Validate` functions and register them in the webhook registry.
+
+### Alternative Designs Considered
+
+#### Plugin model for hypervisor backend integration
+
+The alternative design that we evaluated was a plugin model for hypervisor integration. In this approach, KubeVirt would define a set of core interfaces (similar to the ones described above), but instead of implementing these interfaces in-tree, KubeVirt would provide the mechanism for dynamically loading external implementations at runtime. Each hypervisor backend (e.g., KVM, MSHV, or future hypervisors) could supply its own plugin, packaged and maintained in a separate repository. These plugins would register with KubeVirt components (virt-launcher, virt-handler, etc.) through a well-defined contract.
+
+While the plugin-based approach offers strong decoupling and extensibility, it requires significant refactoring of the KubeVirt codebase. Hypervisor-specific logic is currently invoked from multiple components (e.g., virt-launcher, virt-handler, API), and introducing a dynamic plugin mechanism would involve redesigning these interactions and adding lifecycle management for external modules.
+
+For the initial implementation of multi-hypervisor support, we chose an in-tree design to achieve a working solution faster. This approach allows us to validate the abstraction layer, experiment with real-world scenarios, and identify what works and what does not. With these learnings, we will be better positioned to propose a robust plugin-based architecture for multi-hypervisor support in the future.
+
+#### Alternatives to KubeVirt CR API change
+
+This VEP proposes to add the `HypervisorConfiguration` field to the `KubevirtConfiguration` CRD. The rationale behind this choice was to allow the cluster admin to declare outright which hypervisor they want KubeVirt to target. Furthermore, it follows the `ArchConfiguration` field in `KubevirtConfiguration` CRD. In addition, we also considered the following alternatives to configure KubeVirt to target a specific hypervisor:
+
+- **Use a separate CRD**: Devise a `KubeVirtHypervisorConfiguration` CRD. If it is found in the `kubevirt` namespace, use it. If not, fall back to `KVM`. This is an approach that meets our requirements, but we prefer to add the `HypervisorConfiguration` field following the precedent of `ArchConfiguration`.
```

Comment:
I find this reasoning a bit weak.
There are several advantages to choose this one, including the ability to change the API easier, have a more granular permission scheme, deployability and a better fit for the future out-of-tree extension.

What are the cons of going this way?